controllers/noderefutil/indexer.go

```diff
@@ -65,14 +66,46 @@ func IndexNodeByProviderID(o client.Object) []string {
 		panic(fmt.Sprintf("Expected a Node but got a %T", o))
 	}
 
-	if node.Spec.ProviderID != "" {
-		providerID, err := NewProviderID(node.Spec.ProviderID)
-		if err != nil {
-			// Failed to create providerID, skipping.
-			return nil
-		}
-		return []string{providerID.IndexKey()}
+	if node.Spec.ProviderID == "" {
+		return nil
+	}
+
+	providerID, err := NewProviderID(node.Spec.ProviderID)
+	if err != nil {
+		// Failed to create providerID, skipping.
+		return nil
+	}
+	return []string{providerID.IndexKey()}
+}
+
+// AddMachineProviderIDIndex adds the machine providerID index to the
+// managers cache.
+func AddMachineProviderIDIndex(ctx context.Context, mgr ctrl.Manager) error {
```

Comment:
Is this the best package to place this function? cc @CecileRobertMichon

Reply:
What folks think about having an `index` package under `api/v1alpha4/` where we can define all related indexes? Indexes use specific paths which are version-aware.

Reply:
cc @ykakarap @sbueringer @fabriziopandini (given that this relates to ClusterClass work as well)

Reply:
They are version-aware based on that the fields we use to calculate can change between versions, right? 

Overall I would try to keep the api package relatively lightweight if possible. But if those are utils which we **need** versioned, I guess that's currently the best place to put them

Reply:
+1
We already went through a similar discussion in the part when this index started to be used by more that one controller; having a dedicated package makes sense to me.

only one clarification. it is required to implement the /index package only for the current API version, right?

Reply:
The `clusterv1.MachineProviderIDIndex` which is API dependant is already in the API package.
`AddMachineProviderIDIndex` is not and it should not IMO, that would increase the controller-runtime surface over the API package which is a dependency burden for consumers wanting to bring the API https://github.com/kubernetes-sigs/cluster-api/issues/3465

Reply:
@enxebre Would it be if it's a dedicated sub-package? I would have guessed that users usually just import the "main" `api/v1alpha4` package and then don't get the dependencies of e.g. `api/v1alpha4/index`

Reply:
That would work I think. The dependency is enforced by webhooks nevertheless.

Reply:
Ref: https://github.com/kubernetes-sigs/cluster-api/pull/5004